include/logging/log_msg2.h

```diff
@@ -34,6 +34,23 @@ typedef uint64_t log_timestamp_t;
 typedef uint32_t log_timestamp_t;
 #endif
 
+/* Determine if auto/__auto_type is supported. If not then runtime approach must always
+ * be used.
+ *
+ * Supported by:
+ * - C++ (auto)
+ * - GCC 4.9.0 https://gcc.gnu.org/gcc-4.9/changes.html
+ * - Clang 3.8
+ */
+#if defined(__cplusplus)
+#define Z_AUTO_TYPE auto
+#elif ((__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__) >= 40900) || \
+	((__clang_major__ * 10000 + __clang_minor__ * 100 + __clang_patchlevel__) >= 30800)
+#define Z_AUTO_TYPE __auto_type
+#else
+#error "Compiler lacks support for __auto_type. Enable CONFIG_LOG2_ALWAYS_RUNTIME."
```

Comment:
should be

#ifndef CONFIG_LOG2_ALWAYS_RUNTIME
#error
#endif

right?

I see that below in this file we are checking if this define is set, so I assume it is not always unset.

Reply:
"move fast and break things" ;)

looks like a good job overall, at least the parts I was able
to understand.